File: view/src/main/java/com/github/lany192/text/EllipsizeTextView.java
package com.github.lany192.text;

import android.content.Context;
import android.graphics.Color;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.URLSpan;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.github.lany192.view.R;
import com.hjq.toast.ToastUtils;

public class EllipsizeTextView extends BoxTextView {
    private final String moreText = "...全文";

    public EllipsizeTextView(Context context) {
        this(context, null);
    }

    public EllipsizeTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public EllipsizeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setMovementMethod(LinkMovementMethod.getInstance());
        setHighlightColor(Color.TRANSPARENT);
    }

    @Override
    public void setText(CharSequence text, TextView.BufferType type) {
        if (TextUtils.isEmpty(text)) {
            text = "";
        }
        //添加超链接点击跳转功能
        SpannableStringBuilder builder = new SpannableStringBuilder(text);
        URLSpan[] spans = builder.getSpans(0, text.length(), URLSpan.class);
        for (URLSpan span : spans) {
            int startIndex = builder.getSpanStart(span);
            int endIndex = builder.getSpanEnd(span);
            builder.setSpan(new ClickableSpan() {

                @Override
                public void updateDrawState(TextPaint ds) {
                    ds.setColor(getResources().getColor(R.color.primary));
                    ds.setUnderlineText(false);
                }

                @Override
                public void onClick(@NonNull View widget) {
                    String title = builder.subSequence(startIndex, endIndex).toString();
                    log.i("点击超链接：" + title + span.getURL());
                    ToastUtils.show("点击超链接：" + title + span.getURL());
                }
            }, startIndex, endIndex, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.removeSpan(span);
        }
        super.setText(builder, type);
    }

    private int getTextNeedWidth(String text) {
        float[] widths = new float[text.length()];
        getPaint().getTextWidths(text, widths);
        int nowLength = 0;
        for (float width : widths) {
            nowLength += width;
        }
        return nowLength;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        if (getLineCount() > getMaxLines()) {
            setText(clipText());
        }
    }

    private SpannableStringBuilder clipText() {
        int indexEnd = getLayout().getLineEnd(getMaxLines() - 1);
        CharSequence temp = getText().subSequence(0, indexEnd);

        int moreWidth = (int) Math.ceil(getTextNeedWidth(moreText));

        int offsetWidth = getTextNeedWidth(String.valueOf(temp.charAt(indexEnd - 1)));
        int countEmoji = 0;
        int offset = 1;
        while (indexEnd > offset && offsetWidth <= moreWidth) {
            boolean isEmoji = isEmoji(temp.charAt(indexEnd - offset));
            if (isEmoji) {
                countEmoji += 1;
            }
            offset++;
            if (indexEnd > offset) {
                if (countEmoji % 2 == 0) {
                    char charText = temp.charAt(indexEnd - offset);
                    offsetWidth += getTextNeedWidth(String.valueOf(charText));
                    if (offsetWidth > moreWidth && isEmoji(charText)) {
                        offset++;
                    }
                }
            } else {
                char charText = temp.charAt(indexEnd - offset);
                offsetWidth += getTextNeedWidth(String.valueOf(charText));
            }
            log.i("测试indexEnd:" + indexEnd);
            log.i("测试offset:" + offset);
            log.i("测试offsetWidth:" + offsetWidth);
            log.i("测试moreWidth:" + moreWidth);
        }
        CharSequence clip = temp.subSequence(0, indexEnd - offset);


        SpannableString sb = new SpannableString(moreText);
        sb.setSpan(new ForegroundColorSpan(Color.BLUE), 3, sb.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
        sb.setSpan(new ClickableSpan() {

            @Override
            public void updateDrawState(@NonNull TextPaint ds) {
                ds.setColor(getResources().getColor(R.color.primary));
                ds.setUnderlineText(false);
            }

            @Override
            public void onClick(@NonNull View widget) {
                ToastUtils.show("点击了全文");
            }
        }, 3, sb.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);

        SpannableStringBuilder stringBuilder = new SpannableStringBuilder(clip);
        stringBuilder.append(sb);
        return stringBuilder;
    }

    private boolean isEmoji(char str) {
        return str == 0x0 || str == 0x9 || str == 0xA
                || str == 0xD || str >= 0x20 && str <= 0xD7FF
                || str >= 0xE000 && str <= 0xFFFD;
    }
}

File: view/src/main/java/com/github/lany192/text/EllipsizeTextView.kt
package com.github.lany192.text

import android.content.Context
import android.graphics.Color
import android.text.*
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.ForegroundColorSpan
import android.text.style.URLSpan
import android.util.AttributeSet
import android.view.View
import android.widget.Toast
import kotlin.math.ceil

open class EllipsizeTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : BoxTextView(context, attrs, defStyleAttr) {

    private val moreText = "...全文"

    init {
        movementMethod = LinkMovementMethod.getInstance()
        highlightColor = Color.TRANSPARENT
    }

    override fun setText(content: CharSequence, type: BufferType?) {
        val text = if (TextUtils.isEmpty(content)) {
            SpannableStringBuilder("")
        } else {
            SpannableStringBuilder(content)
        }
        //添加超链接点击跳转功能
        val spans = text.getSpans(0, text.length, URLSpan::class.java)
        val builder = SpannableStringBuilder(text)
        for (span in spans) {
            builder.setSpan(object : ClickableSpan() {

                override fun updateDrawState(ds: TextPaint) {
                    ds.color = Color.RED
                    ds.isUnderlineText = false
                }

                override fun onClick(widget: View) {
                    Toast.makeText(context, "测试1", Toast.LENGTH_SHORT).show()
                }
            }, text.getSpanStart(span), text.getSpanEnd(span), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        }
        super.setText(builder, type)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        if (lineCount > maxLines) {
            text = clipText()
        }
    }

    private fun clipText(): SpannableStringBuilder {

        val indexEnd = layout.getLineEnd(maxLines - 1)
        val tempText = text.subSequence(0, indexEnd)

        val moreWidth = ceil(layout.paint.measureText(moreText).toDouble()).toInt()

        var offsetWidth = layout.paint.measureText(tempText[indexEnd - 1].toString()).toInt()
        //表情字节个数
        var countEmoji = 0
        var offset = 1
        while (indexEnd > offset && offsetWidth <= moreWidth) {
            //当前字节是否位表情
            val isEmoji = isEmoji(tempText[indexEnd - offset])
            if (isEmoji) {
                countEmoji += 1
            }
            offset++

            if (indexEnd > offset) {
                val text = tempText[indexEnd - offset - 1].toString().trim()
                if (text.isNotEmpty() && countEmoji % 2 == 0) {
                    val charText = tempText[indexEnd - offset]
                    offsetWidth += layout.paint.measureText(charText.toString()).toInt()
                    if (offsetWidth > moreWidth && isEmoji(charText)) {
                        offset++
                    }
                }
            } else {
                val charText = tempText[indexEnd - offset]
                offsetWidth += layout.paint.measureText(charText.toString()).toInt()
            }
        }
        val ssbShrink = tempText.subSequence(0, indexEnd - offset)
        val stringBuilder = SpannableStringBuilder(ssbShrink)


        val sb = SpannableString(moreText)
        sb.setSpan(ForegroundColorSpan(Color.BLUE), 3, sb.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        sb.setSpan(object : ClickableSpan() {
            override fun updateDrawState(paint: TextPaint) {
                paint.color = Color.GREEN
                paint.isUnderlineText = false
            }

            override fun onClick(widget: View) {
                Toast.makeText(context, "点击了全文", Toast.LENGTH_SHORT).show()
            }
        }, 3, sb.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        stringBuilder.append(sb)
        return stringBuilder
    }

    private fun isEmoji(codePoint: Char): Boolean {
        return !(codePoint.toInt() == 0x0 || codePoint.toInt() == 0x9 || codePoint.toInt() == 0xA || codePoint.toInt() == 0xD || codePoint.toInt() in 0x20..0xD7FF || codePoint.toInt() in 0xE000..0xFFFD || codePoint.toInt() in 0x10000..0x10FFFF)
    }
}